### ISSUE TITLE
Query by type does not accept type constraints [DATAJPA-1257]

### ISSUE THREAD

**Issue by spring-projects-issues** [spring-projects/spring-data-jpa#1591]
**[Thomas Lang](https://jira.spring.io/secure/ViewProfile.jspa?name=thomas.lang@th-deg.de)** opened **[DATAJPA-1257](https://jira.spring.io/browse/DATAJPA-1257?redirect=false)** and commented

Hey there spring data jpa folks,

I don´t know if this is really a bug.
The stackoverflow link tells basically everything you need.

Here the details:
I have an abstract class with others derived from that.
I use a repository to query the resulting database table for the discriminator type.

When i use this repository:
Please note: the class constraint on the query type

```java
/**
     * Finds the given role by a given discriminator value and a certain application user id.
     *
     * @param abstractRole      a given abstract role
     * @param applicationUserId a given application user id
     * @return AbstractRole
     */
    @Query("select ar from AbstractRole ar where type(ar) = ?1 and ar.applicationUser.applicationUserId = ?2")
    AbstractRole findsProjectedByRoleTypeAndApplicationUserId(Class<? extends AbstractRole> abstractRole, long applicationUserId);
```

With the following code:

```java
final AbstractRole abstractRole = abstractRoleRepository.findsProjectedByRoleTypeAndApplicationUserId(AdminRole.class, 6074);
```

I face the following exception:

```java
org.springframework.dao.InvalidDataAccessApiUsageException: Parameter value [6074] did not match expected type [java.lang.Class (n/a)]; nested exception is java.lang.IllegalArgumentException: Parameter value [6074] did not match expected type [java.lang.Class (n/a)]
```

When i use the solution suggested in the stackoverflow link (leave out the type constraint) it works as expected.

```java
/**
     * Finds the given role by a given discriminator value and a certain application user id.
     *
     * @param abstractRole      a given abstract role
     * @param applicationUserId a given application user id
     * @return AbstractRole
     */
    @Query("select ar from AbstractRole ar where type(ar) = ?1 and ar.applicationUser.applicationUserId = ?2")
    AbstractRole findsProjectedByRoleTypeAndApplicationUserId(Class<?> abstractRole, long applicationUserId);
```

So basically the constraint would be cool to have, because it makes the usage safer.
But for now, i can life with that.

Please respond if you need some more information about this.

Have a nice day.

By the way:
Keep up the good job with spring data jpa. This framework is outstanding.
Makes me happy to have this project on board for my daily business.

Regards from Lower Bavaria!
Thomas

---

**Affects:** 1.11.10 (Ingalls SR10)

**Reference URL:** https://stackoverflow.com/questions/47052706/spring-data-repository-query-by-class-with-generic-class

**Issue Links:**
- [DATAJPA-1497](https://jira.spring.io/browse/DATAJPA-1497) Allow to add a special parameter type from a Spring Data JPA extension

3 votes, 6 watchers

**Comment by nseb** [comment]
Hello , the problem is still relevant ? Because the .class in hql query was deprecated and now we need use type(myclass) and pass a class in parameter. This is going to get really blocking

**Comment by mp911de** [comment]
This has been addressed via spring-projects/spring-data-commons#2770
